### ISSUE TITLE
Bug(linter): some global objects are not included

### ISSUE THREAD

**Issue by cijiugechu** [oxc-project/oxc#732]
#### Description
Currently, the `no-undef` rule treats `console` as an undefined global variable, this results in an unnecessary diagnostic message.

#### Reproduction
```js
console.log(1);
```

#### Possible cause
Currently, oxc linter only uses [BUILTINS](https://github.com/web-infra-dev/oxc/blob/main/crates/oxc_linter/src/globals.rs#L8) to detect global variables, which misses detection of certain runtimes, such as `browser` and `node.js`, where global objects such as console are included. Does oxc need a config to specify runtime like eslint?

**Comment by Boshen** [comment]
Let's port https://eslint.org/docs/latest/use/configure/language-options#specifying-environments

**Comment by simonxabris** [comment]
Hey! I'd be interested in taking the crack at this, I just need a bit more info so I don't go down the wrong path from the start.
 
My initial understanding based on the description and the comment was, that the `env` option is supposed to be parsed out from the `eslint` config, which is already read [on this line](https://github.com/web-infra-dev/oxc/blob/main/crates/oxc_linter/src/lib.rs#L112C8-L112C32), but then I realised that it's never used anywhere, so that's probably not the way to go.

Should a new `env` cli argument be added, which users can set and then based on the value of that, [BUILTINS](https://github.com/web-infra-dev/oxc/blob/main/crates/oxc_linter/src/globals.rs#L8) should be dynamically created with the appropriate values?

**Comment by Boshen** [comment]
> Hey! I'd be interested in taking the crack at this, I just need a bit more info so I don't go down the wrong path from the start.
> 
> My initial understanding based on the description and the comment was, that the `env` option is supposed to be parsed out from the `eslint` config, which is already read [on this line](https://github.com/web-infra-dev/oxc/blob/main/crates/oxc_linter/src/lib.rs#L112C8-L112C32), but then I realised that it's never used anywhere, so that's probably not the way to go.
> 
> Should a new `env` cli argument be added, which users can set and then based on the value of that, [BUILTINS](https://github.com/web-infra-dev/oxc/blob/main/crates/oxc_linter/src/globals.rs#L8) should be dynamically created with the appropriate values?

@simonxabris It seems like the `--env` cli flag https://eslint.org/docs/latest/use/command-line-interface#--env is the thing that we want to port from eslint. All behaviours should match exactly to that of eslint.

**Comment by camc314** [comment]
To add, I think we should also port 

https://github.com/typescript-eslint/typescript-eslint/blob/1ed0733c7fea0dcb4f10dc0a54a4dafde57511df/packages/scope-manager/src/lib/

as just porting the eslint config will still mean that using a type e.g. `ReturnType` will still complain

**Comment by Boshen** [comment]
> just porting the eslint config will still mean that using a type e.g. ReturnType will still complain

Can you elaborate on this? 

> I think we should also port

This file looks crazy lol. We already have a scope manager, what usecases do we need to add?

**Comment by camc314** [comment]
> Can you elaborate on this?


e.g.
```
  ⚠ eslint(no-undef): Disallow the use of undeclared variables
    ╭─[../../<PATH>.tsx:11:1]
 11 │
 12 │ export function ExampleTable({ data }: { data: ReturnType<typeof useData>['data'] }) {
    ·                                                ──────────
 13 │     const columns: ColumnDef<Data>[] = [
    ╰────
  help: 'ReturnType' is not defined.
  ```

I think, even if we ported the eslint configurations, these errors would still be reported

> This file looks crazy lol. We already have a scope manager, what usecases do we need to add?

i think we just need to add those items to the scope (mainly types) (depending on env specified)
